File: linear_motor.py
from gpiozero import OutputDevice
import time

# ピン番号の定義 (BCMモード = GPIO番号で指定)
IN1_PIN = 14  # GPIO 14
IN2_PIN = 15  # GPIO 15

# setup() の代わりに、ここでオブジェクトを作成します
# これでピンが自動的に出力モードに設定されます
in1 = OutputDevice(IN1_PIN)
in2 = OutputDevice(IN2_PIN)

def motor_control():
    """モーター制御のシーケンス"""
    
    # 正転
    print("正転 (Forward)")
    in1.on()   # RPi.GPIOの High
    in2.off()  # RPi.GPIOの Low
    time.sleep(2)

    # ブレーキ
    print("ブレーキ (Brake)")
    in1.on()
    in2.on()
    time.sleep(2)

    # 逆転
    print("逆転 (Backward)")
    in1.off()
    in2.on()
    time.sleep(2)

    # 停止（惰性）
    print("停止 (Coast)")
    in1.off()
    in2.off()
    time.sleep(2)

def main():
    """メイン処理"""
    try:
        while True:  # Arduinoのloop()に相当
            motor_control()
            
    except KeyboardInterrupt:
        # Ctrl+C でプログラムを終了したとき
        print("\nプログラムを終了します。")
        
    finally:
        # RPi.GPIO.cleanup() の代わりに、デバイスを閉じる
        # （実際には try ブロックの外に出た時点で自動でクリーンアップされますが
        #  明示的に書いておくと確実です）
        in1.close()
        in2.close()
        print("GPIOクリーンアップ完了")

# このスクリプトが直接実行された場合にmain()関数を呼び出す
if __name__ == "__main__":
    main()

File: linear_motor.cpp
#include <gpiod.hpp>
#include <iostream>
#include <unistd.h>

const std::string CHIP_NAME = "gpiochip4";
const int IN1_PIN = 14; 
const int IN2_PIN = 15; 

int main() {
    try {
        gpiod::chip chip(CHIP_NAME);
        gpiod::line line_in1 = chip.get_line(IN1_PIN);
        gpiod::line line_in2 = chip.get_line(IN2_PIN);

        // 出力として要求 (初期値 LOW)
        line_in1.request({"test-motor", gpiod::line_request::DIRECTION_OUTPUT, 0}, 0);
        line_in2.request({"test-motor", gpiod::line_request::DIRECTION_OUTPUT, 0}, 0);
        std::cout << "GPIO lines acquired. Starting test." << std::endl;

        // 順方向 (extend_motor に相当)
        std::cout << "Forward (IN1=1, IN2=0)" << std::endl;
        line_in1.set_value(1);
        line_in2.set_value(0);
        sleep(2);

        // 逆方向 (shrink_motor に相当)
        std::cout << "Backward (IN1=0, IN2=1)" << std::endl;
        line_in1.set_value(0);
        line_in2.set_value(1);
        sleep(2);

        // 停止 (stop_motor に相当)
        std::cout << "Stop (IN1=0, IN2=0)" << std::endl;
        line_in1.set_value(0);
        line_in2.set_value(0);
        sleep(1);

    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << std::endl;
        return 1;
    }
    std::cout << "Test finished." << std::endl;
    return 0;
}
